### PR TITLE
Add valve component

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

Adds support for valve components/entities

Related PRs:
- aioesphomeapi: https://github.com/esphome/aioesphomeapi/pull/852
- Home Assistant: https://github.com/home-assistant/core/pull/115341
- Home Assistant docs: https://github.com/home-assistant/home-assistant.io/pull/32441#event-12583125844
- Home Assistant (bump aioesphomeapi): https://github.com/home-assistant/core/pull/115729

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3763

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/web_server/web_server.cpp [hunk 1]
```diff
@@ -330,6 +330,13 @@ void WebServer::handle_index_request(AsyncWebServerRequest *request) {
   }
 #endif
 
+#ifdef USE_VALVE
+  for (auto *obj : App.get_valves()) {
+    if (this->include_internal_ || !obj->is_internal())
+      write_row(stream, obj, "valve", "<button>Open</button><button>Close</button>");
+  }
+#endif
+
```
Comment (jesserockz) [inline esphome/components/web_server/web_server.cpp]: I would not worry about updating v1 webserver stuff for new entities (I didnt add the date/time stuff)

File: esphome/components/valve/valve.h [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  ESPDEPRECATED("close() is deprecated, use make_call().set_command_close() instead.", "2021.9")
+  void close();
+  /** Stop the valve.
+   *
+   * This is a legacy method and may be removed later, please use `.make_call()` instead.
+   * As per solution from issue #2885 the call should include perform()
+   */
+  ESPDEPRECATED("stop() is deprecated, use make_call().set_command_stop().perform() instead.", "2021.9")
+  void stop();
```
Comment (jesserockz) [inline esphome/components/valve/valve.h]: Doesnt make sense to add deprecated functions on a new section of code =)

File: esphome/components/valve/valve.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+namespace esphome {
+namespace valve {
+
+const extern float VALVE_OPEN;
+const extern float VALVE_CLOSED;
+
+#define LOG_VALVE(prefix, type, obj) \
+  if ((obj) != nullptr) { \
+    ESP_LOGCONFIG(TAG, "%s%s '%s'", prefix, LOG_STR_LITERAL(type), (obj)->get_name().c_str()); \
+    auto traits_ = (obj) -> get_traits(); \
```
Comment (kbx81) [inline esphome/components/valve/valve.h]: Haha, why does it complain about this one line? 😅 
```suggestion
    auto traits_ = (obj)->get_traits(); \
```

File: esphome/components/template/valve/__init__.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+
+async def to_code(config):
+    var = await valve.new_valve(config)
+    await cg.register_component(var, config)
+    if CONF_LAMBDA in config:
+        template_ = await cg.process_lambda(
+            config[CONF_LAMBDA], [], return_type=cg.optional.template(float)
+        )
+        cg.add(var.set_state_lambda(template_))
```
Comment (jesserockz) [inline esphome/components/template/valve/__init__.py]: More walruses are required in this file =)
```suggestion
    if lambda_config := config.get(CONF_LAMBDA):
        template_ = await cg.process_lambda(
            lambda_config, [], return_type=cg.optional.template(float)
        )
        cg.add(var.set_state_lambda(template_))
```

File: esphome/components/template/valve/__init__.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        cv.Optional(CONF_ASSUMED_STATE, default=False): cv.boolean,
+        cv.Optional(CONF_HAS_POSITION, default=False): cv.boolean,
+        cv.Optional(CONF_OPEN_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_CLOSE_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_STOP_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_TOGGLE_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_POSITION_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_RESTORE_MODE, default="RESTORE"): cv.enum(
+            RESTORE_MODES, upper=True
+        ),
```
Comment (jesserockz) [inline esphome/components/template/valve/__init__.py]: I think a while back we came to a decision that everything should not restore by default and it needs to be user choice.
```suggestion
        cv.Optional(CONF_RESTORE_MODE, default="NO_RESTORE"): cv.enum(
            RESTORE_MODES, upper=True
        ),
```

File: esphome/components/template/valve/template_valve.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  void dump_config() override;
+
+  float get_setup_priority() const override;
+
+ protected:
+  void control(const valve::ValveCall &call) override;
+  valve::ValveTraits get_traits() override;
+  void stop_prev_trigger_();
+
+  TemplateValveRestoreMode restore_mode_{VALVE_RESTORE};
```
Comment (jesserockz) [inline esphome/components/template/valve/template_valve.h]:
```suggestion
  TemplateValveRestoreMode restore_mode_{VALVE_NO_RESTORE};
```

File: esphome/components/valve/__init__.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        ),
+    }
+)
+
+
+async def setup_valve_core_(var, config):
+    await setup_entity(var, config)
+
+    if CONF_DEVICE_CLASS in config:
+        cg.add(var.set_device_class(config[CONF_DEVICE_CLASS]))
```
Comment (jesserockz) [inline esphome/components/valve/__init__.py]:
```suggestion
    if device_class_config := config.get(CONF_DEVICE_CLASS):
        cg.add(var.set_device_class(device_class_config))
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/template/valve/__init__.py (hunk 4 above)
Comment (kbx81) [reply esphome/components/template/valve/__init__.py]: Hahaha, we talked about this and I completely forgot to circle back and do it. 🤦🏻‍♂️  Will fix.

**Comment by nagyrobi** [comment]
The Endstop and Time Based valves (copied functionality from the coresponding cover types) are the next ones needed...

File: esphome/components/template/valve/__init__.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    ),
+)
+async def valve_template_publish_to_code(config, action_id, template_arg, args):
+    paren = await cg.get_variable(config[CONF_ID])
+    var = cg.new_Pvariable(action_id, template_arg, paren)
+    if state_config := config.get(CONF_STATE):
+        template_ = await cg.templatable(state_config, args, float)
+        cg.add(var.set_position(template_))
+    if position_config := config.get(CONF_POSITION):
```
Comment (jesserockz) [inline esphome/components/template/valve/__init__.py]: Gotta be careful with config values that can be falsy
```suggestion
    if (position_config := config.get(CONF_POSITION)) is not None:
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**